File: vuln-sandbox/java/cwe-22/PathTraversalSafe.java
// SAFE: java-path-file-concat — canonical path check prevents directory traversal
// Rule: java-path-file-concat | CWE-22 | Expected: TrueNegative

import javax.servlet.http.HttpServlet;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.OutputStream;

public class PathTraversalSafe extends HttpServlet {

    private static final String UPLOAD_DIR = "/var/app/uploads";

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) {
        String filename = request.getParameter("file");

        try {
            // SAFE: resolve canonical path and verify it stays within the allowed directory
            File uploadDir = new File(UPLOAD_DIR).getCanonicalFile();
            File file = new File(uploadDir, filename).getCanonicalFile();

            if (!file.getPath().startsWith(uploadDir.getPath() + File.separator)) {
                response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Invalid file path");
                return;
            }

            if (!file.exists() || !file.isFile()) {
                response.sendError(HttpServletResponse.SC_NOT_FOUND, "File not found");
                return;
            }

            response.setContentType("application/octet-stream");
            OutputStream out = response.getOutputStream();

            try (FileInputStream fis = new FileInputStream(file)) {
                byte[] buffer = new byte[4096];
                int bytesRead;
                while ((bytesRead = fis.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}

File: vuln-sandbox/rust/cwe-22/path_traversal_safe.rs
// SAFE: rust-fs-read-format — path canonicalization and containment check prevent traversal
// Rule: rust-fs-read-format | CWE-22 | Expected: TrueNegative

use std::fs;
use std::path::{Path, PathBuf};

fn read_user_file(base_dir: &Path, user_filename: &str) -> Result<Vec<u8>, String> {
    let base = base_dir.canonicalize().map_err(|e| e.to_string())?;

    // SAFE: resolve the full path and verify it stays within the allowed directory
    let file_path = base.join(user_filename);
    let canonical = file_path.canonicalize().map_err(|_| "File not found".to_string())?;

    if !canonical.starts_with(&base) {
        return Err("Path traversal detected".to_string());
    }

    fs::read(&canonical).map_err(|e| e.to_string())
}

fn main() {
    let base_dir = PathBuf::from("/uploads");
    let filename = std::env::args().nth(1).unwrap_or_default();

    match read_user_file(&base_dir, &filename) {
        Ok(content) => println!("read {} bytes", content.len()),
        Err(e) => eprintln!("Error: {}", e),
    }
}
